### PR TITLE
[Driver][SYCL] Do not set default device code split mode for FPGA

### PR DESCRIPTION
**Pull request by qichaogu**
The Driver passes default device code split mode (-split=auto) to
sycl-post-link when -fsycl-device-code-split is not set. This should
not be applied for FPGA, because FPGA is using SYCL_EXTERNAL for
some of their library configurations but auto device code split mode
impacts this behavior.

Signed-off-by: Qichao Gu <qichao.gu@intel.com>

### CODE REVIEW

File: clang/test/Driver/sycl-offload-with-split.c [hunk 1]
```diff
@@ -296,6 +296,16 @@
 // RUN:    | FileCheck %s -check-prefixes=CHK-NO-SPLIT
 // CHK-NO-SPLIT-NOT: sycl-post-link{{.*}} -split{{.*}}
 
+// Check no device code split mode is passed to sycl-post-link when -fsycl-device-code-split is not set but -fintelfpga is set
+// RUN:   %clang -### -fsycl -fintelfpga %s 2>&1 | FileCheck %s -check-prefixes=CHK-NO-SPLIT
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload-with-split.c:300]: A couple suggestions to generalize the test/test description:
```suggestion
// Check no device code split mode is passed to sycl-post-link when -fsycl-device-code-split is not set but the FPGA triple is set
// RUN:   %clang -### -fsycl -fsycl-targets=spir64_fpga-unknown-unknown-sycldevice %s 2>&1 | FileCheck %s -check-prefixes=CHK-NO-SPLIT
```

File: clang/test/Driver/sycl-offload-with-split.c [hunk 2]
```diff
@@ -296,6 +296,16 @@
...
+
+// Check device code split mode is honored when both -fsycl-device-code-split and -fintelfpga are set
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload-with-split.c:302]:
```suggestion
// Check device code split mode is honored when both -fsycl-device-code-split and the FPGA triple are set
```

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -8919,7 +8919,7 @@ void SYCLPostLink::ConstructJob(Compilation &C, const JobAction &JA,
     else
       // split must be off
       assert(StringRef(A->getValue()) == "off");
-  } else {
+  } else if (getToolChain().getTriple().getArchName() != "spir64_fpga") {
     // auto is the default split mode
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:8923]:
```suggestion
    // auto is the default split mode, unless FPGA is targeted - in that case,
    // split is off by default.
```

**Review by AGindinson — COMMENTED** [review]
> This should not be applied for FPGA.

Could you please extend the description with a brief reasoning block?

File: clang/test/Driver/sycl-offload-with-split.c (hunk 1 above)
Comment (qichaogu) [reply clang/test/Driver/sycl-offload-with-split.c]: Fixed. Thanks for the suggestion.

File: clang/test/Driver/sycl-offload-with-split.c (hunk 2 above)
Comment (qichaogu) [reply clang/test/Driver/sycl-offload-with-split.c]: Fixed.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 3 above)
Comment (qichaogu) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Fixed.

**Comment by qichaogu** [comment]
> > This should not be applied for FPGA.
> 
> Could you please extend the description with a brief reasoning block?

Please see update in the commit description:
The Driver passes default device code split mode (-split=auto) to
sycl-post-link when -fsycl-device-code-split is not set. This should
not be applied for FPGA, because FPGA is using SYCL_EXTERNAL for
some of their library configurations but auto device code split mode
impacts this behavior.

**Review by AGindinson — DISMISSED** [review]
LGTM, thanks!

File: clang/test/Driver/sycl-offload-with-split.c [hunk 4]
```diff
@@ -296,6 +296,16 @@
 // RUN:    | FileCheck %s -check-prefixes=CHK-NO-SPLIT
 // CHK-NO-SPLIT-NOT: sycl-post-link{{.*}} -split{{.*}}
 
+// Check no device code split mode is passed to sycl-post-link when -fsycl-device-code-split is not set and the target is FPGA
+// RUN:   %clang -### -fsycl -fsycl-targets=spir64_fpga-unknown-unknown-sycldevice %s 2>&1 | FileCheck %s -check-prefixes=CHK-NO-SPLIT
```
Comment (mlychkov) [inline clang/test/Driver/sycl-offload-with-split.c:300]: Maybe skip sycldevice triple part if this is going to be dropped soon?

Comment (bader) [reply clang/test/Driver/sycl-offload-with-split.c:300]: It's already dropped.
https://github.com/intel/llvm/commit/29a1f167060e69d4d90d1e6a15d4998fd4556c4d

Comment (qichaogu) [reply clang/test/Driver/sycl-offload-with-split.c]: Removed 'sycldevice' from triple.

**Review by mlychkov — APPROVED** [review]

**Review by AGindinson — APPROVED** [review]

**Comment by bader** [comment]
@qichaogu, this patch breaks https://github.com/intel/llvm-test-suite/blob/intel/SYCL/AOT/multiple-devices.cpp. Please, take a look at the logs.

**Review by bader — CHANGES_REQUESTED** [review]
@qichaogu, please, resolve merge conflict.

**Comment by qichaogu** [comment]
> @qichaogu, please, resolve merge conflict.

We need to put this pull request on hold, because it has dependency on runtime support.

**Comment by bader** [comment]
> We need to put this pull request on hold, because it has dependency on runtime support.

I converted it to draft to indicate that it's still "work in progress".

**Pull request closed without merging**